README.md
# analyze-restful-api-load-test-results

![Maintained](https://img.shields.io/maintenance/yes/2022.svg?style=flat)
[![MIT license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)
[![Python 3.9](https://img.shields.io/badge/python-3.9-blue.svg)](https://www.python.org/downloads/release/python-390/)
[![Requirements Status](https://requires.io/github/simonsdave/analyze-restful-api-load-test-results/requirements.svg?branch=master)](https://requires.io/github/simonsdave/analyze-restful-api-load-test-results/requirements/?branch=master)
[![CodeFactor](https://www.codefactor.io/repository/github/simonsdave/analyze-restful-api-load-test-results/badge/master)](https://www.codefactor.io/repository/github/simonsdave/analyze-restful-api-load-test-results/overview/master)
[![CircleCI](https://circleci.com/gh/simonsdave/analyze-restful-api-load-test-results/tree/master.svg?style=shield)](https://circleci.com/gh/simonsdave/analyze-restful-api-load-test-results/tree/master)
[![codecov](https://codecov.io/gh/simonsdave/analyze-restful-api-load-test-results/branch/master/graph/badge.svg)](https://codecov.io/gh/simonsdave/analyze-restful-api-load-test-results)
[![docker-simonsdave/analyze-restful-api-load-test-results](https://img.shields.io/badge/docker-simonsdave%2Fanalyze--restful--api--load--test--results-blue.svg)](https://hub.docker.com/r/simonsdave/analyze-restful-api-load-test-results/)

As part of a microservice's CI process it's ideal to
load test a microservice's RESTful API to confirm
that changes to the microservice's codebase do not negatively
affect the microservice's performance characteristics.
There are a number of very good load testing tools - for
example [locust](http://locust.io),
[JMeter](http://jmeter.apache.org)
and [k6](https://k6.io).
These tools can easily drive load into a microservice.
The real challenge is the automated assessment of the load test results
and this is where ```analyze-restful-api-load-test-results``` helps.

## Process Overview

* drive load through the microservice using your favorite load
  testing tool and produce a [TSV](https://en.wikipedia.org/wiki/Tab-separated_values)
  file containing timing and success for each request issued during the
  load test
* in the context of a CI pipeline, ```analyze-restful-api-load-test-results```
  can be configured to numerically analyze the microservice's performance
  and produce a failing exit code if the performance falls outside of acceptable
  bounds
* run ```analyze-restful-api-load-test-results``` can optionally
  be configured to graphically assess the microservice's performance - saving
  the graphical assessment as a build artifact can prove useful when attempting
  to assess a numerical assessment failure

## Input File Format

The [TSV](https://en.wikipedia.org/wiki/Tab-separated_values) file is
input for ```analyze-restful-api-load-test-results```.
The file should be UTF8 encoded
and contain the fields below separated by a single tab.
[This](samples/001-input.tsv) is a sample input file.

* timestamp
* request type - typically this is going to be an HTTP verb but
  it can be any non-blank string
* success flag - 1 = success; 0 = failure
* client identifier - a non-blank string - only used to enable ```analyze-restful-api-load-test-results```
  to calculate concurrency levels - for [locust](http://locust.io) it's typically
  a locust ID and for [k6](https://k6.io) it's a VU id
* response time in milliseconds - a floating point number

## Running ```analyze-restful-api-load-test-results```

```analyze-restful-api-load-test-results``` is packaged in a docker image.
Why? [NumPy](http://www.numpy.org) and [Matplotlib](http://matplotlib.org)
are used by ```analyze-restful-api-load-test-results```
for numeric and graphical assessment of a microservice's performance.
Both these packages take a long time to install.
A CI pipeline should execute in the minimum time possible
and installing [NumPy](http://www.numpy.org) and [Matplotlib](http://matplotlib.org)
would unacceptably increase CI pipeline execution time.
Packaging ```analyze-restful-api-load-test-results``` in a docker image
means the CI pipeline only has to download the docker image which is
generally quite quick.

The following command runs ```analyze-restful-api-load-test-results```'s
numerical analysis.

```bash
~> docker container run \
    --rm \
    -i \
    simonsdave/analyze-restful-api-load-test-results:latest \
    analyze-restful-api-load-test-results.sh \
    < samples/001-input.tsv
=====================================================================================
21,426 @ 25 from 2017-07-19 01:25:02.410082+00:00 to 2017-07-19 01:30:02.245437+00:00
=====================================================================================

Request Type                 Ok Error         m        b      Min       50       95       99      Max
-----------------------------------------------------------------------------------------------------
GET                       17017     0    0.0551      293      106      290      432      509      842
PUT                        4409     0    0.0351      531      257      508      765     1131     1748

=====================================================================================
~>
```

The following command runs ```analyze-restful-api-load-test-results```'s
numerical and graphical analysis.
The graphical analysis creates a PDF doc called ```/tmp/dave.pdf```
which can be copied out of the container using ```docker container cp```.

```bash
~> docker container run \
    -name graphs
    -i \
    simonsdave/analyze-restful-api-load-test-results:latest \
    analyze-restful-api-load-test-results.sh \
    --graphs=/tmp/graphs.pdf \
    < samples/001-input.tsv
=====================================================================================
21,426 @ 25 from 2017-07-19 01:25:02.410082+00:00 to 2017-07-19 01:30:02.245437+00:00
=====================================================================================

Request Type                 Ok Error         m        b      Min       50       95       99      Max
-----------------------------------------------------------------------------------------------------
GET                       17017     0    0.0551      293      106      290      432      509      842
PUT                        4409     0    0.0351      531      257      508      765     1131     1748

=====================================================================================
~> docker container cp graphs:/tmp/graphs.pdf .
~> docker container rm graphs
~>
```

The graphical analysis in ```dave.pdf``` will contain 1 page per request type.
Each page will contain a single graph that looks like the graph below.
A sample of the full analysis can be found [here](samples/001-input.pdf).

![](samples/001-input.png)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
Format of this file follows [these](http://keepachangelog.com/) guidelines.
This project adheres to [Semantic Versioning](http://semver.org/).

## [%RELEASE_VERSION%] - [%RELEASE_DATE%]

### Added

- [CodeFactor](https://www.codefactor.io) badge to main README.md
- add [clair-cicd](https://github.com/simonsdave/clair-cicd) docker image vulnerability
  assessment to CircleCI pipeline
- [LGTM](https://lgtm.com/) badges to main README.md
- add CircleCI docker executor [authenticated pull](https://circleci.com/docs/2.0/private-images/)
- per [this](https://discuss.circleci.com/t/old-linux-machine-image-remote-docker-deprecation/37572) article, added
  explicit version to ```setup_remote_docker``` in CircleCI pipeline
- added resource_class to CircleCI config

### Changed

- ```dev-env``` v0.5.14 -> v0.6.21
- ```matplotlib``` 2.2.3 -> 3.6.1
- ```numpy``` 1.15.3 -> 1.23.3
- ```python-dateutil``` 2.8.0 -> 2.8.1
- added markdown and json linting to CircleCI pipeline
- added README.rst building to CircleCI pipeline
- added 2022 to LICENSE
- CircleCI setup_remote_docker version 19.03.13 -> 20.10.17

### Removed

- Travis CI
- removed LGTM from main README.md


analyze_restful_api_load_test_results/__init__.py
import datetime
import optparse
import re
import sys

import dateutil.parser
import numpy
from matplotlib.backends.backend_pdf import PdfPages
import matplotlib.pyplot as plt


__version__ = '0.9.0'


class CommandLineParser(optparse.OptionParser):

    def __init__(self):
        optparse.OptionParser.__init__(
            self,
            'usage: %prog [options]',
            description='This utility analyzes load test results')

        default = 0.1
        help_msg = 'max slope for all requests by type - default = %.2f' % default
        self.add_option(
            '--max-slope',
            action='store',
            dest='max_slope',
            default=default,
            type='float',
            help=help_msg)

        default = '/dev/null'
        help_msg = 'pdf filename for graphs - default = %s' % default
        self.add_option(
            '--graphs',
            action='store',
            dest='graphs',
            default=default,
            type='string',
            help=help_msg)

    def parse_args(self, *args, **kwargs):
        (clo, cla) = optparse.OptionParser.parse_args(self, *args, **kwargs)
        if 0 != len(cla):
            self.error('invalid command line args')
        return (clo, cla)


class Response(object):

    def __init__(self, request_type, timestamp, success, vu, response_time):
        object.__init__(self)

        self.request_type = request_type
        self.timestamp = timestamp
        self.success = success
        # k6 calls them virtual users (vu) and gives each vu a unique integer identifier
        # locust calls them locusts and gives each one a unique identifier
        self.vu = vu
        self.response_time = response_time

    def get_seconds_since_start(self, responses):
        return (self.timestamp - responses.first_timestamp).total_seconds()

    def _get_bucket_size_in_seconds(self, responses):
        total_number_seconds_in_test = (responses.last_timestamp - responses.first_timestamp).total_seconds()
        total_number_of_buckets_so_things_look_ok = 100
        return total_number_seconds_in_test / total_number_of_buckets_so_things_look_ok

    def get_seconds_since_start_bucket(self, responses):
        seconds_since_start = int(round(self.get_seconds_since_start(responses), 0))
        return seconds_since_start - (seconds_since_start % self._get_bucket_size_in_seconds(responses))


class Responses(object):

    def __init__(self):
        object.__init__(self)

        self.responses = []
        self.responses_by_request_type = {}
        self.responses_by_vus = {}
        self.first_timestamp = datetime.datetime(2990, 1, 1).replace(tzinfo=dateutil.tz.tzutc())
        self.last_timestamp = datetime.datetime(1990, 1, 1).replace(tzinfo=dateutil.tz.tzutc())

    def add(self, response):
        self.responses.append(response)

        if response.request_type not in self.responses_by_request_type:
            self.responses_by_request_type[response.request_type] = []
        self.responses_by_request_type[response.request_type].append(response)

        if response.vu not in self.responses_by_vus:
            self.responses_by_vus[response.vu] = []
        self.responses_by_vus[response.vu].append(response)

        self.first_timestamp = min(response.timestamp, self.first_timestamp)
        self.last_timestamp = max(response.timestamp, self.last_timestamp)

    def request_types(self):
        return [request_type for request_type in self.responses_by_request_type.keys()]

    def responses_for_request_type(self, request_type):
        return self.responses_by_request_type[request_type]

    def successes_for_request_type(self, request_type):
        return [response for response in self.responses_by_request_type[request_type] if response.success]

    def failures_for_request_type(self, request_type):
        return [response for response in self.responses_by_request_type[request_type] if not response.success]

    def number_of_vus(self):
        return len(self.responses_by_vus.keys())

    def __len__(self):
        return len(self.responses)


class Main(object):

    def load_data(self):
        responses = Responses()

        reg_ex_pattern = (
            r'^\s*'
            r'(?P<timestamp>.+)\t'
            r'(?P<request_type>.+)\t'
            r'(?P<success>\d{1})\t'
            r'(?P<vu>.+)\t'
            r'(?P<response_time>\d*\.\d+|\d+)'
            r'\s*$'
        )
        reg_ex = re.compile(reg_ex_pattern)

        line_number = 0
        for line in sys.stdin:
            line_number += 1
            try:
                match = reg_ex.match(line)
                if match:
                    timestamp = dateutil.parser.parse(match.group('timestamp'))
                    if not timestamp.tzinfo:
                        timestamp = timestamp.replace(tzinfo=dateutil.tz.tzlocal())
                    request_type = match.group('request_type')
                    success = int(match.group('success'))
                    vu = match.group('vu')
                    response_time = float(match.group('response_time'))

                    response = Response(request_type, timestamp, success, vu, response_time)
                    responses.add(response)
                else:
                    print('ERROR - %d: invalid input format - %s' % (line_number, line.strip()))
            except Exception as ex:
                print('ERROR - %d: %s' % (line_number, ex))
                print('>>>%s<<<' % line)

        return responses

    def numerical_analysis(self, responses, max_slope):
        """Returns 0 on success and 1 on failure."""
        overall_title = '%s @ %s from %s to %s' % (
            '{:,}'.format(len(responses)),
            '{:,}'.format(responses.number_of_vus()),
            responses.first_timestamp,
            responses.last_timestamp,
        )
        line = '=' * len(overall_title)
        print(line)
        print(overall_title)
        print(line)
        print('')

        percentiles = [50, 95, 99]
        fmt = '%-25s %5d %5d %9.4f' + '%9.0f' * (2 + len(percentiles) + 1)
        request_types = responses.request_types()
        request_types.sort()

        title_fmt = '%-25s %5s %5s ' + '%9s' * (3 + len(percentiles) + 1)
        args = [
            'Request Type',
            'Ok',
            'Error',
            'm',
            'b',
            'Min',
        ]
        args.extend(percentiles)
        args.append('Max')
        title = title_fmt % tuple(args)
        print(title)
        print('-' * len(title))

        return_value = 0

        for request_type in request_types:
            responses_for_request_type = responses.responses_for_request_type(request_type)

            seconds_since_start = [
                response.get_seconds_since_start(responses) for response in responses_for_request_type
            ]

            response_times = [response.response_time for response in responses_for_request_type]

            degree_of_the_fitting_polynomial = 1
            m, b = numpy.polyfit(seconds_since_start, response_times, degree_of_the_fitting_polynomial)
            args = [
                request_type,
                len(responses.successes_for_request_type(request_type)),
                len(responses.failures_for_request_type(request_type)),
                m,
                b,
                min(response_times),
            ]
            args.extend(numpy.percentile(numpy.array(response_times), percentiles))
            args.append(max(response_times))
            print(fmt % tuple(args))

            if max_slope < abs(m):
                return_value = 1

        print('')
        print('=' * len(overall_title))

        return return_value

    def generate_graphs(self, responses, graphs):
        with PdfPages(graphs) as pdf:
            request_types = responses.request_types()
            request_types.sort()

            for request_type in request_types:
                responses_for_request_type = responses.responses_for_request_type(request_type)

                response_times_in_buckets = {}
                for response in responses_for_request_type:
                    seconds_since_start_bucket = response.get_seconds_since_start_bucket(responses)
                    if seconds_since_start_bucket not in response_times_in_buckets:
                        response_times_in_buckets[seconds_since_start_bucket] = []
                    response_times_in_buckets[seconds_since_start_bucket].append(response.response_time)

                xs = [x for x in response_times_in_buckets.keys()]
                xs.sort()

                tabloid_width = 17
                tabloid_height = 11
                plt.figure(figsize=(tabloid_width, tabloid_height))

                column_labels = ['m', 'b']
                row_labels = []
                row_colours = []
                cells = []

                m_fmt = '%.4f'
                b_fmt = '%.0f'

                ys = [max(response_times_in_buckets.get(x, [0])) for x in xs]
                m, b = numpy.polyfit(xs, ys, 1)
                cells.append([m_fmt % m, b_fmt % b])
                row_labels.append('max')
                row_colours.append('yellow')
                plt.plot(xs, ys, row_colours[-1], label=row_labels[-1], zorder=1)

                percentiles = [99, 90]
                percentile_colours = ['orange', 'red']
                zorders = [2, 3]
                for percentile, percentile_colour, zorder in zip(percentiles, percentile_colours, zorders):
                    ys = [numpy.percentile(response_times_in_buckets.get(x, [0]), percentile) for x in xs]
                    m, b = numpy.polyfit(xs, ys, 1)
                    cells.append([m_fmt % m, b_fmt % b])
                    # see TeX markup @ https://matplotlib.org/users/mathtext.html
                    # for how the superscripting works
                    row_labels.append('$%d^{th} percentile$' % percentile)
                    row_colours.append(percentile_colour)
                    plt.plot(
                        xs,
                        ys,
                        row_colours[-1],
                        label=row_labels[-1],
                        zorder=zorder)

                ys = [min(response_times_in_buckets.get(x, [0])) for x in xs]
                m, b = numpy.polyfit(xs, ys, 1)
                cells.append([m_fmt % m, b_fmt % b])
                row_labels.append('min')
                row_colours.append('blue')
                plt.plot(xs, ys, row_colours[-1], label=row_labels[-1], zorder=4)

                # table of analysis results will also act as legend
                plt.table(
                    colWidths=[0.1] * 3,
                    cellText=cells,
                    rowLabels=row_labels,
                    rowColours=row_colours,
                    rowLoc='right',
                    colLabels=column_labels,
                    loc='bottom')
                # bbox=[0, -0.5, 1, 0.275])

                plt.grid(True)

                plt.xlabel(
                    'Seconds Since Test Start',
                    fontsize='large',
                    fontweight='bold')
                plt.ylabel(
                    'Response Time\n(milliseconds)',
                    fontsize='large',
                    fontweight='bold')

                hours, remainder = divmod((responses.last_timestamp - responses.first_timestamp).total_seconds(), 3600)
                minutes, _ = divmod(remainder, 60)
                title = '%s\n(%.0f%% of %s requests @ %d concurrency for %d hours %d minutes)\n' % (
                    request_type.replace('-', ' '),
                    (len(responses) * 100.0) / len(responses),
                    '{:,}'.format(len(responses)),
                    responses.number_of_vus(),
                    hours,
                    minutes)
                plt.title(
                    title,
                    fontsize='xx-large',
                    fontweight='bold')
                pdf.savefig()
                plt.close()
